### ISSUE TITLE
Arguments, variables, and shelling out at the top level

### ISSUE THREAD

**Issue by casey** [casey/just#2]
Now that `make` is gone, there are some useful features that `j` no longer supports.

This `justfile` uses variables to cut down on boilerplate and improve readability. It also uses the `device` variable to let the user configure which device should be used:

```
nexus5  = 3d8569a787ad6
nexus7  = 44f60ceb5f75b
nexus6p = e1d267ee8d22d
nexus5x = 31e82532ef348
s7      = 3dbc633f55736

device = $(s7)

adb = adb -s $(device)
app = buz.fizbuz.android.dev

grant: 
    $(adb) shell pm grant $(app) android.permission.ACCESS_FINE_LOCATION
    $(adb) shell pm grant $(app) android.permission.CAMERA
    $(adb) shell pm grant $(app) android.permission.GET_ACCOUNTS
    $(adb) shell pm grant $(app) android.permission.WRITE_EXTERNAL_STORAGE
    $(adb) shell pm grant $(app) android.permission.READ_EXTERNAL_STORAGE
```

The `compile` recipe below uses arguments passed from the command line to select which subprojects to invoke compile in:

```
compile:
    for ARG in $$ARGS; do cd $$ARG; just compile ; cd $$WORKSPACE; done
```

And finally, this `justfile` fragment uses `make`'s `$(shell ...)` syntax to set a local variable to the result of a shell command:

```
localhost = $(shell dumpinterfaces | cut -d: -f2 | sed 's/\/.*//' | sed 's/ //g')
```

To summarize, there are three desirable features here:
1. Allowing variables to be set and used in multiple recipes
2. Allowing arguments to be passed in on the command line
3. Allowing variables to be set to the output of a command invocation

These features should probably be designed together, since they'll all affect the same areas of a `justfile`'s syntax. The syntax for setting variables at the top level will affect 1 and 2, and the syntax for variable substitution will be shared by all of them.

**Comment by casey** [comment]
For passing arguments on the command line, I think the old syntax is good.

This calls the bar and baz recipes with three arguments, `a`, `b`, and `c`:

`j bar baz -- a b c`

I've been thinking of some kind of more complex syntax for grouping recipes with arguments, for example:

`j bar [ a b c ] baz [ x y z ]`
or
`j [bar a b c] [baz x y z]`

But I think that's probably overkill. Having whatever follows a `--` be interpreted differently is a well established pattern for command line utilities, and I think that it will be intuitive for users to grasp.

Also, if you want call multiple recipes with different arguments, you could just write a recipe that takes all the arguments and invokes other recipes with them.

I would also like to extend label syntax to note how many arguments a recipe expects. For example, if the `baz` recipe takes two arguments, a `device` and a `path`, then it would appear in the justfile like so:

```
baz server path:
  curl http://$(server):8080/routes/$(path)
```

This way, the we can also print a useful error message if the user supplies the wrong number of arguments at the command line.

We can support variadic arguments like so:

```
# 0 or more arguments
foo *servers:
  echo $(servers)

# 1 or more arguments
foo +servers:
  echo $(servers)
```

**Comment by casey** [comment]
On to variable substitution!

In the above examples I used the make syntax for variable substitution: `$(variable-name)`, which I'm not sure is ideal, since it conflicts with shell syntax.

One possibility is to use a template inspired syntax:

```
var = hello

foo:
  echo {{hello}}

foo-python:
  #!/usr/bin/env python
  print("{{hello}}")
```

This will be familiar to template programmers, and isn't likely to conflict with shell or language syntax.

We could also do direct text replacement:

```
VAR = hello

foo:
  echo VAR
```

Since the text of recipes is present in the justfile, they can just be sure to pick uppercase variable names which don't appear in any command.

**Comment by casey** [comment]
And on to setting variables at the top level and shelling out!

This concerns how variables are set at the top level.

```
# allow relaxed sh syntax, with spaces around '='
foo = bar

# only allow sh syntax, no spaces
bar=frobny

# don't allow barewords:
baz = "hello"

# shell out with ``
bob = `echo hello`

# shell out with $()
frank = $(echo hello)
```

I really don't have any strong notions of what to do here.

**Comment by casey** [comment]
@aoeu What are your thoughts on the right syntax for all this stuff?

**Comment by casey** [comment]
One note about arguments. `make` actually supports the `a b c:` syntax, but it just means that the recipe has three names, `a`, `b`, and `c`, and you can run it with any of those names. I'm going to assume that that's not useful for `j`, but I could be wrong.

**Comment by casey** [comment]
This is all done!

Variables, arguments, and saving the result of shell invocations are all implemented.
